lib/DxilContainer/DxilContainerReader.cpp

```diff
@@ -1,93 +1,86 @@
-          reader.GetStateObjectConfig_Flags());
-        break;
-      case DXIL::SubobjectKind::LocalRootSignature:
-        bLocalRS = true;
-      case DXIL::SubobjectKind::GlobalRootSignature: {
-        const void *pOutBytes;
-        uint32_t OutSizeInBytes;
-        if (!reader.GetRootSignature(&pOutBytes, &OutSizeInBytes)) {
-          result = false;
-          continue;
-        }
-        subobjects.CreateRootSignature(reader.GetName(), bLocalRS, pOutBytes, OutSizeInBytes);
-        break;
-      }
-      case DXIL::SubobjectKind::SubobjectToExportsAssociation: {
-        uint32_t NumExports = reader.GetSubobjectToExportsAssociation_NumExports();
-        std::vector<llvm::StringRef> Exports;
-        Exports.resize(NumExports);
-        for (unsigned i = 0; i < NumExports; ++i) {
-          Exports[i] = reader.GetSubobjectToExportsAssociation_Export(i);
-        }
-        subobjects.CreateSubobjectToExportsAssociation(reader.GetName(),
-          reader.GetSubobjectToExportsAssociation_Subobject(),
-          Exports.data(), NumExports);
-        break;
-      }
-      case DXIL::SubobjectKind::RaytracingShaderConfig:
-        subobjects.CreateRaytracingShaderConfig(reader.GetName(),
-          reader.GetRaytracingShaderConfig_MaxPayloadSizeInBytes(),
-          reader.GetRaytracingShaderConfig_MaxAttributeSizeInBytes());
-        break;
-      case DXIL::SubobjectKind::RaytracingPipelineConfig:
-        subobjects.CreateRaytracingPipelineConfig(reader.GetName(),
-          reader.GetRaytracingPipelineConfig_MaxTraceRecursionDepth());
-        break;
-      case DXIL::SubobjectKind::HitGroup:
-        subobjects.CreateHitGroup(reader.GetName(),
-          reader.GetHitGroup_Type(),
-          reader.GetHitGroup_AnyHit(),
-          reader.GetHitGroup_ClosestHit(),
-          reader.GetHitGroup_Intersection());
-          break;
-      case DXIL::SubobjectKind::RaytracingPipelineConfig1:
-        subobjects.CreateRaytracingPipelineConfig1(
-            reader.GetName(),
-            reader.GetRaytracingPipelineConfig1_MaxTraceRecursionDepth(),
-            reader.GetRaytracingPipelineConfig1_Flags());
-        break;
-      }
-    } catch (hlsl::Exception &) {
-      result = false;
-    }
+HRESULT DxilContainerReader::Load(_In_ const void* pContainer, _In_ uint32_t containerSizeInBytes) {
+  if (pContainer == nullptr) {
+    return E_FAIL;
+  }
+
+  const DxilContainerHeader *pHeader = IsDxilContainerLike(pContainer, containerSizeInBytes);
+  if (pHeader == nullptr) {
+    return E_FAIL;
+  }
+  if (!IsValidDxilContainer(pHeader, containerSizeInBytes)) {
+    return E_FAIL;
   }
-  return result;
+
+  m_pContainer = pContainer;
+  m_uContainerSize = containerSizeInBytes;
+  m_pHeader = pHeader;
+  
+  return S_OK;
 }
 
+HRESULT DxilContainerReader::GetVersion(_Out_ DxilContainerVersion *pResult) {
+  if (pResult == nullptr) return E_POINTER;
+  if (!IsLoaded()) return E_NOT_VALID_STATE;
```

Comment:
This too will need to be added to WinAdapter.h:

`#define E_NOT_VALID_STATE 0x8007139F`